### PR TITLE
Zahar

### PR DESCRIPTION
**Pull request by ZaraGiraffe**
I have made the first try to compete the clap pipeline

### CODE REVIEW

File: src/data/tts.py [hunk 1]
```diff
@@ -35,20 +34,6 @@ def generate(self, text: str, speaker: str) -> np.array:
         return np.random.randn(16_000 * 10)
 
 
-class VitsTTSProvider(BaseTTSProvider):
```
Comment (anakib1) [inline src/data/tts.py:38]: What reason to delete existing providers?

File: src/data/tts.py [hunk 2]
```diff
@@ -161,3 +84,41 @@ def generate(self, text: str, speaker: str = "alloy") -> np.array:
...
+        self.speaker_mapper = SpeakerMapper(speakers_bank=self.model.speakers)
+
+    def generate(self, text: str, speaker: str) -> np.array:
+        if not text:
+            return
+        speaker = self.speaker_mapper.map(speaker)
+        audio = self.model.apply_tts(text=text,
+                                speaker=speaker,
+                                sample_rate=self.silero_rate)
+        audio = librosa.resample(audio.numpy(), orig_sr=self.silero_rate, target_sr=self.gen_rate)
```
Comment (anakib1) [inline src/data/tts.py]: maybe torch.Resample

File: src/models/audio.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            self,
+            proj_dim: int,
+            model_dim: int = 384,
+            checkpoint: str = "openai/whisper-tiny",
+            device: str = "cpu"
+    ):
+        super().__init__(proj_dim, device)
+        self.model = WhisperModel.from_pretrained(checkpoint).encoder.to(self.device)
+        self.extractor = WhisperFeatureExtractor.from_pretrained(checkpoint)
+        self.projection_layer = torch.nn.Linear(model_dim, self.proj_dim).ti(self.device)
```
Comment (anakib1) [inline src/models/audio.py]: ti? 😸

File: src/models/audio.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        feats = self.extractor(
+            audios=audio,
+            sampling_rate=16000,
+            return_tensors="pt",
+        )
+        feats = {k: v.to(self.device) for k, v in feats.items()}
+        encoder_output = self.model(**feats).last_hidden_state
+        mask = get_mask(feats["attention_mask"], encoder_output.shape[1])
+        embeddings = masked_emb_pooling(mask, encoder_output)
+        if isinstance(audio, np.ndarray):
```
Comment (anakib1) [inline src/models/audio.py:76]: what does it do

File: src/models/clap.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    def forward(self, texts: Union[List[str], str], audios: Union[List[np.array], np.array]) -> Dict[str, torch.Tensor]:
+        """
+        :param texts: texts for the input
+        :param audios: audios for the
+        :return: dictionary with cosine distance between embeddings
+            if len(texts) == len(audios) also returns the loss
+        """
+        num_texts = 1 if isinstance(texts, str) else len(texts)
+        num_audios = 1 if isinstance(audios, str) else len(audios)
```
Comment (anakib1) [inline src/models/clap.py]: better do something like it in other method. Like we can have predict() that wraps object in the required shapes but forward should have normal sizes at start

File: src/models/clap.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        self.audio_encoder = audio_encoder
+        self.text_encoder = text_encoder
+        self.loss_fn = torch.nn.CrossEntropyLoss()
+
+    def forward(self, texts: Union[List[str], str], audios: Union[List[np.array], np.array]) -> Dict[str, torch.Tensor]:
+        """
+        :param texts: texts for the input
+        :param audios: audios for the
+        :return: dictionary with cosine distance between embeddings
+            if len(texts) == len(audios) also returns the loss
```
Comment (anakib1) [inline src/models/clap.py]: better have loss always, but it can be None sometimes. Even better torch.NaN

File: src/models/clap.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        """
+        num_texts = 1 if isinstance(texts, str) else len(texts)
+        num_audios = 1 if isinstance(audios, str) else len(audios)
+        texts_emb = self.text_encoder(texts)
+        audio_emb = self.audio_encoder(audios)
+        logits = texts_emb @ torch.transpose(audio_emb, 0, 1)
+        if num_texts == 1 or num_audios == 1:
+            logits = logits.squeeze()
+        ans = {"logits": logits}
+        if num_texts == num_audios and num_texts != 1:
```
Comment (anakib1) [inline src/models/clap.py]: why we need num_texts == num_audios

File: src/models/audio.py (hunk 3 above)
Comment (ZaraGiraffe) [reply src/models/audio.py]: Lol, that happened because I added .to(device) to WhisperEncoder after i have checked

File: src/data/tts.py (hunk 2 above)
Comment (ZaraGiraffe) [reply src/data/tts.py]: Yeah, that sounds cleaner. I will change, but I will not be surprised if it uses librosa internally

File: src/data/tts.py (hunk 1 above)
Comment (ZaraGiraffe) [reply src/data/tts.py:38]: I do not remember, but I guess I tried it and it was terrible even in english tts

File: src/models/audio.py (hunk 4 above)
Comment (ZaraGiraffe) [reply src/models/audio.py:76]: there are documentation in get_mask and masked_emb_pooling  functions, but in a few words it transforms attention mask from feature extractor (so that the dims are correct) and use this mask to mask and pool embeddings

File: src/models/clap.py (hunk 5 above)
Comment (ZaraGiraffe) [reply src/models/clap.py]: I thought about it, I can do it if you want, but I am not sure whether this will be usefull. HF processor for example can handle multiple inputs and it is convinient, that is why I did, what I have done

File: src/models/clap.py (hunk 6 above)
Comment (ZaraGiraffe) [reply src/models/clap.py]: Hmm, maybe you are right, I will do it

File: src/models/clap.py (hunk 7 above)
Comment (ZaraGiraffe) [reply src/models/clap.py]: because otherwise you will have to have multiple audions generated from one theme, that creates some problems.
Also, I've checked, in the clip paper they trained with len(image_emb) == len(text_emb)

File: src/data/tts.py (hunk 1 above)
Comment (anakib1) [reply src/data/tts.py:38]: let's just leave it, no need to delete working code

File: src/data/tts.py (hunk 2 above)
Comment (anakib1) [reply src/data/tts.py]: no torch has own implementation

File: src/models/audio.py (hunk 4 above)
Comment (anakib1) [reply src/models/audio.py:76]: No I mean why do we have if is instance(audio, np.ndarray)

File: src/models/clap.py (hunk 5 above)
Comment (anakib1) [reply src/models/clap.py]: yes, it can. But that's processor and not model. I just suspect it could be slow in training because we waste time. AFAIK HuggingFace models expect correct shapes always

File: src/models/clap.py (hunk 6 above)
Comment (anakib1) [reply src/models/clap.py]: Like, KeyError is worse then NoneError

File: src/models/audio.py (hunk 4 above)
Comment (ZaraGiraffe) [reply src/models/audio.py:76]: I have added it in case you want to find embedding of single audio. 
If it is a single audio the encoder gives batch with size 1, so this is basically just squeezing
Should I remove the possibility to add a single audio?
